### PR TITLE
Task/296 attach arconnect session to wallet address

### PR DESCRIPTION
**Pull request by jdaev**
Shows a log out dialog when wallet is switched with arconnect.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -7,15 +7,35 @@ The ArDrive Web App allows a user to log in to securely view, upload and manage
 
 Have any questions? Join the ArDrive Discord channel for support, news and updates. https://discord.gg/ya4hf2H
 
-## Development
+## Setting up the Development Environment
```
Comment (fedellen) [inline README.md:10]: Assuming the merge direction listed in the header is correct, I think we need branch `feat/224-Connect-with-ARConnect` to reintegrate with dev to get these changes out of the diff here

File: lib/app_shell.dart [hunk 2]
```diff
@@ -19,10 +22,19 @@ class AppShell extends StatefulWidget {
 
 class _AppShellState extends State<AppShell> {
   bool _showProfileOverlay = false;
-
+  bool _showWalletSwitchDialog = true;
```
Comment (arielmelendez) [inline lib/app_shell.dart:25]: Is the intent for this to only happen 1 time? If so, you might want to add a comment about that in the area where it's used below.

File: lib/blocs/profile/profile_cubit.dart [hunk 3]
```diff
@@ -34,16 +35,49 @@ class ProfileCubit extends Cubit<ProfileState> {
...
+
+    // ArConnect connected to expected wallet - route to login screen
+    if (await arconnect.checkPermissions() &&
+        profile.walletPublicKey == await arconnect.getPublicKey()) {
+      emit(ProfilePromptLogIn());
+      return;
+    }
+
+    // Unexpected ArConnect state - clean up and route to profile screen
+    await _db.transaction(() async {
```
Comment (arielmelendez) [inline lib/blocs/profile/profile_cubit.dart:64]: Should any of this cleanup take place up above in the other ProfilePromptAdd scenarios?

File: lib/blocs/profile/profile_cubit.dart [hunk 4]
```diff
@@ -34,16 +35,49 @@ class ProfileCubit extends Cubit<ProfileState> {
...
-  Future<void> unlockDefaultProfile(String password, bool isArconnect) async {
+  Future<void> unlockDefaultProfile(
+    String password,
+    ProfileType profileType,
+  ) async {
     emit(ProfileLoggingIn());
 
     final profile = await _profileDao.loadDefaultProfile(password);
 
     if (profile != null) {
```
Comment (arielmelendez) [inline lib/blocs/profile/profile_cubit.dart:80]: Consider emitting ProfilePromptAdd and an early return if profile is null and getting rid of the indentation.

File: lib/blocs/profile/profile_cubit.dart [hunk 5]
```diff
@@ -80,6 +114,12 @@ class ProfileCubit extends Cubit<ProfileState> {
   ///
   /// Works even when the user is not authenticated.
   Future<void> logoutProfile() async {
+    try {
+      await arconnect.disconnect();
+    } catch (e) {
+      print(e);
```
Comment (arielmelendez) [inline lib/blocs/profile/profile_cubit.dart:120]: Are there any risks of leaking info to the logs that could be used by malicious actors? I'm looking into Application Performance Monitoring tools for us to use where we can pipe errors like this for aggregated analysis.

File: lib/components/wallet_switch_dialog.dart [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+          ],
+        ),
+        actions: [
+          TextButton(
+            child: Text('Logout'),
+            onPressed: () {
+              Navigator.pop(context);
+              context.read<ProfileCubit>().logoutProfile();
+              if (fromAuthPage ?? false) {
+                print(fromAuthPage);
```
Comment (arielmelendez) [inline lib/components/wallet_switch_dialog.dart:30]: Clean up unnecessary print

File: lib/components/wallet_switch_dialog.dart [hunk 7]
```diff
@@ -0,0 +1,38 @@
+import 'dart:html';
+
+import 'package:ardrive/blocs/blocs.dart';
+import 'package:ardrive/components/app_dialog.dart';
+import 'package:flutter/material.dart';
+import 'package:flutter_bloc/flutter_bloc.dart';
+
+class WalletSwitchDialog extends StatelessWidget {
+  final bool fromAuthPage;
```
Comment (arielmelendez) [inline lib/components/wallet_switch_dialog.dart:9]: Member variable should have a leading underscore.

File: lib/pages/app_router_delegate.dart [hunk 8]
```diff
@@ -67,6 +67,11 @@ class AppRouterDelegate extends RouterDelegate<AppRoutePath>
             signingIn = false;
             notifyListeners();
           }
+
+          if (state is ProfilePromptAdd) {
+            signingIn = true;
```
Comment (arielmelendez) [inline lib/pages/app_router_delegate.dart:72]: Don't necessarily have to clean this up here, but all the members in this class should have leading underscores.

File: lib/pages/profile_auth/components/profile_auth_add_screen.dart [hunk 9]
```diff
@@ -120,7 +120,11 @@ class ProfileAuthAddScreen extends StatelessWidget {
                       ),
                       const SizedBox(height: 16),
                       TextButton(
-                        child: Text('Change wallet'),
+                        child: context
+                                .read<ProfileAddCubit>()
+                                .isArconnectInstalled()
+                            ? Text('LOG OUT')
```
Comment (arielmelendez) [inline lib/pages/profile_auth/components/profile_auth_add_screen.dart:126]: Do the mocks have "LOG OUT" in all caps?

File: lib/pages/profile_auth/components/profile_auth_fail_screen.dart [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+            ),
+            const SizedBox(height: 32),
+            Text(
+              'Sorry, login failed. Please try again.',
+              textAlign: TextAlign.center,
+              style: Theme.of(context).textTheme.headline6,
+            ),
+            const SizedBox(height: 32),
+            TextButton(
+              child: Text('Login'),
```
Comment (arielmelendez) [inline lib/pages/profile_auth/components/profile_auth_fail_screen.dart:34]: 'Login' is a noun, as in the process of logging in. 'Log In' is a call to action. Which do we want here?

File: lib/pages/profile_auth/components/profile_auth_unlock_screen.dart [hunk 11]
```diff
@@ -6,67 +6,80 @@ import 'package:flutter/material.dart';
...
-                          child: Text(
-                            'Forget wallet and change profile',
-                            textAlign: TextAlign.center,
-                          ),
-                          onPressed: () =>
-                              context.read<ProfileCubit>().logoutProfile(),
+          builder: (context, state) {
+            if (state is ProfileUnlockFailure) {
+              return ProfileAuthFailScreen();
+            } else {
```
Comment (arielmelendez) [inline lib/pages/profile_auth/components/profile_auth_unlock_screen.dart:29]: `else` is redundant. Save an indentation.

File: lib/pages/profile_auth/profile_auth_page.dart [hunk 12]
```diff
@@ -10,7 +14,27 @@ import 'components/profile_auth_onboarding_screen.dart';
...
 import 'components/profile_auth_unlock_screen.dart';
 
-class ProfileAuthPage extends StatelessWidget {
+class ProfileAuthPage extends StatefulWidget {
+  @override
+  _ProfileAuthPageState createState() => _ProfileAuthPageState();
+}
+
+class _ProfileAuthPageState extends State<ProfileAuthPage> {
+  bool _showWalletSwitchDialog = true;
```
Comment (arielmelendez) [inline lib/pages/profile_auth/profile_auth_page.dart:23]: Same comment here about leaving a note about how this is used.

**Review by arielmelendez — CHANGES_REQUESTED** [review]
Left a handful of comments and questions. Thanks!

File: lib/blocs/profile/profile_cubit.dart (hunk 5 above)
Comment (jdaev) [reply lib/blocs/profile/profile_cubit.dart:120]: I don't think so. Not from here at least, this one who just print out dart-js interop errors.

File: lib/blocs/profile/profile_cubit.dart (hunk 3 above)
Comment (jdaev) [reply lib/blocs/profile/profile_cubit.dart:64]: No. The cleanup already happens in the other cases as it is in the logout function. It was added here only for unexpected cases.

File: lib/components/wallet_switch_dialog.dart (hunk 7 above)
Comment (jdaev) [reply lib/components/wallet_switch_dialog.dart:9]: Not here. The convention is different for arguments that are used as a named parameter in the class.

File: lib/pages/app_router_delegate.dart (hunk 8 above)
Comment (jdaev) [reply lib/pages/app_router_delegate.dart:72]: Again no, those are named parameters. Can't use underscores here.

File: lib/pages/profile_auth/components/profile_auth_add_screen.dart (hunk 9 above)
Comment (jdaev) [reply lib/pages/profile_auth/components/profile_auth_add_screen.dart:126]: Mocks?

File: lib/pages/profile_auth/profile_auth_page.dart (hunk 12 above)
Comment (jdaev) [reply lib/pages/profile_auth/profile_auth_page.dart:23]: [Added](https://github.com/ardriveapp/ardrive-web/pull/147/commits/85cf9fa36cfb1d372d82ba474e703560c1dec477)

File: lib/pages/profile_auth/components/profile_auth_fail_screen.dart (hunk 10 above)
Comment (jdaev) [reply lib/pages/profile_auth/components/profile_auth_fail_screen.dart]: Changed to "Log In" [here](https://github.com/ardriveapp/ardrive-web/pull/147/commits/85cf9fa36cfb1d372d82ba474e703560c1dec477)

File: lib/components/wallet_switch_dialog.dart (hunk 6 above)
Comment (jdaev) [reply lib/components/wallet_switch_dialog.dart]: [Added](https://github.com/ardriveapp/ardrive-web/pull/147/commits/85cf9fa36cfb1d372d82ba474e703560c1dec477)

File: lib/app_shell.dart (hunk 2 above)
Comment (jdaev) [reply lib/app_shell.dart:25]: [Added](https://github.com/ardriveapp/ardrive-web/pull/147/commits/85cf9fa36cfb1d372d82ba474e703560c1dec477)

File: lib/blocs/profile/profile_cubit.dart (hunk 4 above)
Comment (jdaev) [reply lib/blocs/profile/profile_cubit.dart]: [Added](https://github.com/ardriveapp/ardrive-web/pull/147/commits/21873aa2ce87fe269b2f842e47ecd6aaa147c779)

File: lib/pages/profile_auth/components/profile_auth_add_screen.dart (hunk 9 above)
Comment (arielmelendez) [reply lib/pages/profile_auth/components/profile_auth_add_screen.dart:126]: Design mock ups from Sketch, etc.

File: lib/models/tables/profiles.moor [hunk 13]
```diff
@@ -4,5 +4,7 @@ CREATE TABLE profiles (
     username TEXT NOT NULL,
 
     encryptedWallet BLOB NOT NULL,
-    keySalt BLOB NOT NULL
+    keySalt BLOB NOT NULL,
+    walletPublicKey TEXT NOT NULL,
+    profileType INTEGER NOT NULL
```
Comment (arielmelendez) [inline lib/models/tables/profiles.moor:9]: nit: missing newline at end of file

**Review by arielmelendez — APPROVED** [review]

**Pull request merged**